File: languages/typescript/extensions/compute/sdk.ts
/**
 * COMPUTE — the in-process SDK surface.
 *
 * Same rule as CONTENT and HISTORY: there is no conformance gate on this face and
 * per the operator that is correct, so **the extension is the instrument** — the
 * handler's own wire-gated path runs THROUGH this class rather than beside it, and
 * so does the H7 evaluator seam. Two consumers, one implementation; if this surface
 * were wrong the `compute` category would say so.
 *
 * **This is also the file that owns §4's override prohibition**, which as of v3.29
 * is ours to enforce and was not before. See {@link assertNotBuiltinOverride}.
 */

import {
  Ecf,
  Entity,
  type ContentStore,
  type EntityTree,
  type codec,
} from "entity-core-protocol-typescript";

import {
  BUILTINS_PREFIX,
  DEFAULT_MAX_DEPTH,
  DEFAULT_MAX_OPS,
} from "./types.js";
import {
  ComputeError,
  emptyScope,
  evaluate,
  isError,
  materialize,
  type Budget,
  type EvalContext,
} from "./internal/evaluator.js";

/** §9.3's two peer-wide defaults, as a value rather than as two arguments. */
export interface EvaluatorLimits {
  readonly maxOperations: number;
  readonly maxDepth: number;
}

export const DEFAULT_LIMITS: EvaluatorLimits = {
  maxOperations: DEFAULT_MAX_OPS,
  maxDepth: DEFAULT_MAX_DEPTH,
};

/**
 * The subset of the peer this extension reads.
 *
 * Declared structurally rather than as `Peer`, because the H7 evaluator seam and the
 * handler path reach the same services from two different objects (`HandlerContext.peer`
 * is `PeerServices`, and the composition holds a `Peer`). Naming the four members we
 * actually use is what lets one implementation serve both without either caller
 * widening what it hands us.
 */
export interface ComputePeerServices {
  readonly tree: EntityTree;
  readonly contentStore: ContentStore;
  readonly localPeerId: string;
}

/**
 * What an evaluation produced. Exactly one of the two fields is non-null.
 *
 * **`value` is the MATERIALIZED form and never an in-flight one** (§2.3, v3.19c
 * option α). This interface is the compute→non-compute crossing for both of this
 * extension's consumers — the handler's `eval` and the H7 evaluator seam — so it is
 * the one place the boundary has to be enforced, and enforcing it here is why
 * neither consumer had to learn what an in-flight value is.
 */
export interface EvalOutcome {
  readonly value: codec.EcfValue | Entity | null;
  readonly error: ComputeError | null;
  /** Steps actually charged — `maxOperations` minus what is left. §4.2's counter. */
  readonly operationsUsed: number;
}

export interface EvaluateOptions {
  /** §3.2's `params.budget` knob. `null` means the caller did not ask. */
  readonly budget?: number | null;
  /** §4.2 Tier 0 — the `content_store_access` allowance. Default false. */
  readonly contentStoreAccess?: boolean;
  /** §4.2 step 1 — the envelope's pre-authorized `included` map, keyed by hex hash. */
  readonly included?: ReadonlyMap<string, Entity>;
  /**
   * §4.2 Tier 2 — an installed subgraph's SEALED set, hex-encoded.
   *
   * Empty unless the caller is §7.2's reactive path, which is the only caller that
   * has one: the set is sealed by §3.3 Phase 2b and lives on the subgraph metadata.
   * An earlier draft of this class passed the ENCOUNTERED set here, which would have
   * made every hash an evaluation happened to write Tier-2 authorized — an
   * authorization tier admitting whatever the program itself produced.
   */
  readonly authorizedDataHashes?: ReadonlySet<string>;
  /**
   * §4.1 / §6.3 — the two path predicates, when the caller HOLDS a capability object.
   *
   * **Default `() => true`, and that default is the declared substrate gap rather
   * than an oversight.** On the explicit-eval path this port has no capability object
   * to narrow with: `HandlerContext.callerCapability` is a token and the peer exposes
   * no path-scope predicate an extension can call — so §6.2 is satisfied at the
   * DISPATCH boundary and not per tree read. See
   * {@link CAPABILITY_CHECK_IS_DISPATCH_SCOPED}.
   *
   * §7.2's reactive path is different in kind and that is why these are parameters:
   * it evaluates under the INSTALLATION GRANT, which is an entity it fetched from the
   * content store, so it can and does answer both per path. One port, two authority
   * models, because the spec gives the two paths two authorities.
   */
  readonly canReadPath?: (path: string) => boolean;
  readonly canWritePath?: (path: string) => boolean;
  /**
   * §3.2 E1 — scope bindings pre-populated by a dispatch layer (`operation`, `params`,
   * `resource`, `caller_capability` for an entity-native body). Empty by default. A binding
   * carries no authority: it is a value an expression can read through `compute/lookup/scope`.
   */
  readonly bindings?: ReadonlyMap<string, codec.EcfValue | Entity>;
}

/**
 * The evaluator, as an in-process object.
 *
 * One instance per evaluation is the intended use — it carries the encountered-hash
 * set, which §4.2 scopes to a single evaluation. Reusing one across evaluations would
 * widen `resolve()`'s reach past what the section allows, which is the kind of change
 * that makes more programs work and is still wrong.
 */
export class ComputeEvaluator {
  readonly #peer: ComputePeerServices;
  readonly #limits: EvaluatorLimits;
  readonly #encountered = new Set<string>();
  readonly #dependencies = new Set<string>();

  constructor(peer: ComputePeerServices, limits: EvaluatorLimits = DEFAULT_LIMITS) {
    this.#peer = peer;
    this.#limits = limits;
  }

  /** Paths §7.1 would register as reactive dependencies. Read after an evaluation. */
  get dependencies(): readonly string[] {
    return [...this.#dependencies];
  }

  /**
   * Evaluate `expression`, with `subgraphRoot` as the base for `relative: true` paths.
   *
   * **`subgraphRoot` is a required parameter and not a default**, because §2.1 gives
   * it two different sources — the expression URI for an explicit eval, the subgraph
   * metadata's `root_expression_path` for a reactive one — and "same value, different
   * source" is exactly the kind of sentence that becomes a bug when one caller is
   * allowed to omit it.
   */
  evaluateAt(expression: Entity, subgraphRoot: string, options: EvaluateOptions = {}): EvalOutcome {
    // §5.2's minimum rule, in the one place it can be applied: the caller's ask and
    // the peer default, whichever is smaller. A caller cannot raise its own ceiling.
    const asked = options.budget ?? null;
    const operations =
      asked === null ? this.#limits.maxOperations : Math.min(asked, this.#limits.maxOperations);

    const budget: Budget = { operations, depth: this.#limits.maxDepth };
    const ctx = this.#context(subgraphRoot, options);

    const scope = emptyScope();
    for (const [name, value] of options.bindings ?? []) scope.bindings.set(name, value);
    const out = evaluate(expression, scope, budget, ctx);
    const used = operations - budget.operations;

    if (isError(out)) {
      // `isError` is kind-based (§4.1), so this catches BOTH an in-flight
      // ComputeError and a compute/error entity that evaluated successfully. The
      // second is why the branch converts rather than casting.
      const error =
        out instanceof ComputeError
          ? out
          : new ComputeError(readErrorCode(out), "evaluated to a stored compute/error");
      return { value: null, error, operationsUsed: used };
    }
    // §2.3 / v3.19c option α — THE BOUNDARY. Everything above this line may be an
    // in-flight typed value; nothing below it is. `materialize` also stores what the
    // bare form references, which is the half of the rule that is easy to drop: a
    // returned entity whose `system/hash` field points at nothing the store has is a
    // reference the caller cannot follow, and no vector for that failure exists.
    return { value: materialize(out, ctx), error: null, operationsUsed: used };
  }

  #context(subgraphRoot: string, options: EvaluateOptions): EvalContext {
    const peer = this.#peer;
    const encountered = this.#encountered;
    const dependencies = this.#dependencies;
    return {
      tree: peer.tree,
      contentStore: peer.contentStore,
      localPeerId: peer.localPeerId,
      subgraphRoot,
      included: options.included ?? new Map(),
      hasContentStoreAccess: options.contentStoreAccess === true,
      // §4.2 TIER 2 — empty on the explicit-eval path and the SEALED set on the
      // reactive one. The default is empty rather than "whatever this evaluation has
      // touched": an earlier draft passed the ENCOUNTERED set, which would have made
      // every hash the program itself produced Tier-2 authorized.
      //
      // `markEncountered` still feeds nothing here, and that is also deliberate:
      // §10.1's encountered-during-read guarantee is scoped to compute-typed
      // entities, which Tier 1 already admits. The set is kept so the call sites
      // match §4.4/§4.3 N6.
      authorizedDataHashes: options.authorizedDataHashes ?? new Set<string>(),
      canReadPath: options.canReadPath ?? (() => true),
      // §6.3's `store` write. A distinct member rather than an alias of
      // `canReadPath` because §6.3 gives the two different authorities — a tree READ
      // rides `ctx.capability` and a `store` WRITE rides the caller's with the
      // handler forbidden from substituting its own grant (no silent escalation) —
      // and §7.2 is the caller that supplies two different answers.
      canWritePath: options.canWritePath ?? (() => true),
      registerDependency: (path: string) => {
        dependencies.add(path);
      },
      markEncountered: (hex: string) => {
        encountered.add(hex);
      },
    };
  }
}

/**
 * §6.2's capability check — **which of this extension's entry points narrows per tree
 * read, and which cannot.**
 *
 * §4.1 guards `compute/lookup/tree` with
 * `check_path_permission("get", path, ctx.capability, "system/tree", ...)`. There are
 * three ways into this evaluator and they do not get the same answer, which is D13's
 * face amendment applied one level down — to the *caller*, not the peer:
 *
 * | entry point | capability object | §6.2 |
 * |---|---|---|
 * | `system/compute:eval` (handler) | `HandlerContext.callerCapability` | **per tree read** |
 * | §7.2 reactive re-evaluation | the installation grant, from the store | **per tree read** |
 * | the H7 evaluator seam | none — `ExpressionRequest` carries the `Execute`, not a verified token | dispatch-scoped |
 *
 * **THIS CONSTANT USED TO READ `true` AND THE REASON IT WAS WRONG IS WORTH KEEPING.**
 * It was `true` because a contract in this tree stated that the peer exposed no
 * path-scope predicate an extension could call. It does:
 * `Permissions.checkPathPermission` is public, takes a `CapabilityToken`, and is the
 * primitive this repo itself routed to keystone as **H9** — closed, landed, and on
 * our own tracker. The gap was never in the substrate; it was a source read nobody
 * re-ran against the tree it described (D12 / L8).
 *
 * It stays exported, and false, because the third row is still true and an inline
 * `() => true` on that path would be the same behaviour with nothing to grep for.
 */
export const CAPABILITY_CHECK_IS_DISPATCH_SCOPED = false;

/**
 * §4's override prohibition — **ours to enforce as of v3.29, and it was not before.**
 *
 * Through v3.28 §3.5 described this rule as *"a subset of"* core's `system/*`
 * reservation and told implementers that enforcing that reservation needed *"no
 * separate compute-specific guard."* `ENTITY-CORE-PROTOCOL` 0.8.2.13 **withdrew the
 * reservation**, so v3.29 restates the prohibition on its own basis: it binds **every
 * installation path** because it is a cross-peer determinism requirement, not a
 * namespace policy. Two peers disagreeing about what `"add"` means is an interop
 * failure — which is precisely what an installed evaluator could cause, since the H7
 * seam hands it every body the built-in path refuses.
 *
 * There is now nothing upstream that refuses a registration at
 * `system/compute/builtins/*` on our behalf, so this function is the enforcement
 * point, and it is called from {@link installCompute}'s path rather than left as
 * documentation.
 */
export function assertNotBuiltinOverride(pattern: string): void {
  const relative = pattern.startsWith("/") ? pattern.slice(pattern.indexOf("/", 1) + 1) : pattern;
  if (relative === BUILTINS_PREFIX || relative.startsWith(BUILTINS_PREFIX + "/")) {
    throw new Error(
      `EXTENSION-COMPUTE §4: a handler MUST NOT be registered at ${pattern}. ` +
        "The builtin handlers are the cross-peer determinism floor; overriding one " +
        "makes two peers disagree about what an operation means. (v3.29 — this rule " +
        "no longer delegates to the withdrawn core system/* reservation.)",
    );
  }
}

function readErrorCode(value: unknown): string {
  // A `compute/error` with no readable `code` is malformed. Saying so with the
  // §9.1 code for a malformed expression beats inventing a plausible one.
  //
  // `unknown` rather than `Entity`, because `isError` is kind-based and admits an
  // in-flight construct whose `entity_type` is `compute/error` as well as a stored
  // one (§2.4 — the disposition must not depend on which form arrived).
  const entity =
    value instanceof Entity
      ? value
      : ((value as { entity?: Entity }).entity ?? null);
  if (entity === null) return "invalid_expression";
  return Ecf.optText(entity.data, "code") ?? "invalid_expression";
}

File: languages/python/extensions/compute/sdk.py
"""COMPUTE — the in-process SDK surface, for the `python` peer.

Same rule as CONTENT and HISTORY: there is no conformance gate on this face and per the
operator that is correct, so **the extension is the instrument** — the handler's own
wire-gated path runs THROUGH this class rather than beside it. Two consumers, one
implementation; if this surface were wrong the `compute` category would say so.

**This is also the file that owns §4's override prohibition**, which as of v3.29 is ours to
enforce and was not before. See :func:`assert_not_builtin_override`.

**THERE IS NO H7 EVALUATOR SEAM ON THIS PEER, so this port has one consumer where
`typescript` has two** — see ``__init__.py``'s ``install_compute``. That removes the one
entry point §6.2 could not narrow per tree read, which is why
:data:`CAPABILITY_CHECK_IS_DISPATCH_SCOPED`'s reasoning is different here even though its
value is the same.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable

from entity_core.peer.model import Entity

from ._internal.evaluator import (
    Budget,
    ComputeError,
    EvalContext,
    as_compute_error,
    empty_scope,
    evaluate,
    is_error,
    materialize,
)
from .types import BUILTINS_PREFIX, DEFAULT_MAX_DEPTH, DEFAULT_MAX_OPS


@dataclass(frozen=True, slots=True)
class EvaluatorLimits:
    """§9.3's two peer-wide defaults, as a value rather than as two arguments."""

    max_operations: int
    max_depth: int


#: §9.3's recommended pair, as the default every entry point falls back to.
DEFAULT_LIMITS = EvaluatorLimits(max_operations=DEFAULT_MAX_OPS, max_depth=DEFAULT_MAX_DEPTH)


@dataclass(frozen=True, slots=True)
class EvalOutcome:
    """What an evaluation produced. Exactly one of the first two fields is non-None.

    **:attr:`value` is the MATERIALIZED form and never an in-flight one** (§2.3, v3.19c
    option α). This is the compute -> non-compute crossing for both of this port's consumers
    — the handler's ``eval`` and §7.2's reactive re-evaluation — so it is the one place the
    boundary has to be enforced, and enforcing it here is why neither consumer had to learn
    what an in-flight value is.

    ``value`` is ``None`` both when the outcome is an error AND when the expression genuinely
    evaluated to null (§4.1's ``if`` with no ``else`` returns one, and §4.5 makes it falsy).
    **Read :attr:`error` to tell the two apart, never :attr:`value`.** `typescript` has the
    same ambiguity for the same reason; on this substrate it is sharper because there is no
    tagged null to distinguish.
    """

    value: Any
    error: ComputeError | None
    #: Steps actually charged — ``max_operations`` minus what is left. §4.2's counter.
    operations_used: int


def _always_true(_path: str) -> bool:
    return True


def _no_dependency(_path: str) -> None:
    return None


class ComputeEvaluator:
    """The evaluator, as an in-process object.

    One instance per evaluation is the intended use — it carries the encountered-hash set,
    which §4.2 scopes to a single evaluation. Reusing one across evaluations would widen
    ``resolve()``'s reach past what the section allows, which is the kind of change that makes
    more programs work and is still wrong.

    Takes the PEER rather than a four-member interface. `typescript` declares a
    ``ComputePeerServices`` structural type because its two callers reach the same services
    from two different objects (a ``Peer`` and a ``PeerServices``); here both callers hold the
    peer itself, and the two members this class reads — ``.store`` and ``.local_peer`` — are
    the same two on every path. Declared in ``[sdk_surface].substrate``.
    """

    __slots__ = ("_peer", "_limits", "_encountered", "_dependencies")

    def __init__(self, peer: Any, limits: EvaluatorLimits = DEFAULT_LIMITS) -> None:
        self._peer = peer
        self._limits = limits
        self._encountered: set[str] = set()
        self._dependencies: list[str] = []

    @property
    def dependencies(self) -> tuple[str, ...]:
        """Paths §7.1 would register as reactive dependencies. Read after an evaluation."""
        return tuple(self._dependencies)

    def evaluate_at(
        self,
        expression: Entity,
        subgraph_root: str,
        *,
        budget: int | None = None,
        content_store_access: bool = False,
        included: dict | None = None,
        authorized_data_hashes: frozenset[str] | None = None,
        can_read_path: Callable[[str], bool] | None = None,
        can_write_path: Callable[[str], bool] | None = None,
    ) -> EvalOutcome:
        """Evaluate ``expression``, with ``subgraph_root`` as the base for ``relative: true``
        paths.

        **``subgraph_root`` is a required positional and not a default**, because §2.1 gives
        it two different sources — the expression URI for an explicit eval, the subgraph
        metadata's ``root_expression_path`` for a reactive one — and "same value, different
        source" is exactly the kind of sentence that becomes a bug when one caller is allowed
        to omit it.

        The keyword arguments are `typescript`'s ``EvaluateOptions`` object, in this language's
        idiom (``[sdk_surface].substrate``). Their meanings, in spec terms:

        ``budget``
            §3.2's ``params.budget`` knob. ``None`` means the caller did not ask, which is
            different from passing the default — §5.2's minimum rule is applied below and a
            caller cannot raise its own ceiling.
        ``content_store_access``
            §4.2 Tier 0. Default False, and it is the door
            ``_internal`` exists to keep shut.
        ``included``
            §4.2 step 1 — the envelope's pre-authorized ``included`` map, keyed by hex.
        ``authorized_data_hashes``
            §4.2 Tier 2, an installed subgraph's SEALED set. Empty unless the caller is
            §7.2's reactive path, which is the only caller that has one: the set is sealed by
            §3.3 Phase 2b and lives on the subgraph metadata. **NOT the encountered set** — an
            earlier `typescript` draft passed that, which would have made every hash the
            program itself produced Tier-2 authorized.
        ``can_read_path`` / ``can_write_path``
            §4.1 / §6.3's two path predicates, when the caller HOLDS a capability. Default
            permissive, which is the declared substrate gap on the explicit-eval path rather
            than an oversight — see :data:`CAPABILITY_CHECK_IS_DISPATCH_SCOPED`. §7.2's
            reactive path supplies both, because it evaluates under the INSTALLATION GRANT,
            which is an entity it fetched from the content store.
        """
        # §5.2's minimum rule, in the one place it can be applied: the caller's ask and the
        # peer default, whichever is smaller.
        operations = (
            self._limits.max_operations if budget is None
            else min(budget, self._limits.max_operations)
        )

        counters = Budget(operations=operations, depth=self._limits.max_depth)
        ctx = self._context(subgraph_root, content_store_access, included, authorized_data_hashes,
                            can_read_path, can_write_path)

        out = evaluate(expression, empty_scope(), counters, ctx)
        used = operations - counters.operations

        if is_error(out):
            # `is_error` is kind-based (§4.1), so this catches BOTH an in-flight ComputeError
            # and a compute/error entity that evaluated successfully. The second is why this
            # converts rather than casts.
            return EvalOutcome(value=None, error=as_compute_error(out), operations_used=used)
        # §2.3 / v3.19c option α — THE BOUNDARY. Everything above this line may be an
        # in-flight typed value; nothing below it is. `materialize` also stores what the bare
        # form references, which is the half of the rule that is easy to drop: a returned
        # entity whose `system/hash` field points at nothing the store has is a reference the
        # caller cannot follow, and no vector for that failure exists.
        return EvalOutcome(value=materialize(out, ctx), error=None, operations_used=used)

    def _context(
        self,
        subgraph_root: str,
        content_store_access: bool,
        included: dict | None,
        authorized_data_hashes: frozenset[str] | None,
        can_read_path: Callable[[str], bool] | None,
        can_write_path: Callable[[str], bool] | None,
    ) -> EvalContext:
        dependencies = self._dependencies
        encountered = self._encountered

        def register_dependency(path: str) -> None:
            if path not in dependencies:
                dependencies.append(path)

        def mark_encountered(hex_: str) -> None:
            encountered.add(hex_)

        return EvalContext(
            store=self._peer.store,
            local_peer=self._peer.local_peer,
            subgraph_root=subgraph_root,
            included=included if included is not None else {},
            has_content_store_access=content_store_access is True,
            # §4.2 TIER 2 — empty on the explicit-eval path and the SEALED set on the reactive
            # one. `mark_encountered` deliberately feeds nothing here: §10.1's
            # encountered-during-read guarantee is scoped to compute-typed entities, which
            # Tier 1 already admits. The set is kept so the call sites match §4.4/§4.3 N6.
            authorized_data_hashes=(
                authorized_data_hashes if authorized_data_hashes is not None else frozenset()
            ),
            can_read_path=can_read_path if can_read_path is not None else _always_true,
            # §6.3's `store` write. A distinct member rather than an alias of `can_read_path`
            # because §6.3 gives the two different authorities — a tree READ rides
            # `ctx.capability` and a `store` WRITE rides the caller's with the handler
            # forbidden from substituting its own grant — and §7.2 supplies two different
            # answers.
            can_write_path=can_write_path if can_write_path is not None else _always_true,
            register_dependency=register_dependency,
            mark_encountered=mark_encountered,
        )


#: §6.2's capability check — **which of this extension's entry points narrows per tree read,
#: and which cannot.**
#:
#: §4.1 guards ``compute/lookup/tree`` with
#: ``check_path_permission("get", path, ctx.capability, "system/tree", ...)``. On `typescript`
#: there are THREE ways into the evaluator and they do not get the same answer; **on this peer
#: there are two, and both narrow per tree read**:
#:
#: ===================================  =============================  ==================
#: entry point                          capability                      §6.2
#: ===================================  =============================  ==================
#: ``system/compute:eval`` (handler)    ``DispatchCtx.caller_cap``      **per tree read**
#: §7.2 reactive re-evaluation          the installation grant          **per tree read**
#: the H7 evaluator seam                *does not exist on this peer*   n/a
#: ===================================  =============================  ==================
#:
#: **So the constant is False here for a DIFFERENT reason than it is False on `typescript`,
#: and stating that is the point of keeping it exported.** There it is False because the third
#: row is real and cannot narrow (``ExpressionRequest`` carries the ``Execute`` and no
#: verified token). Here the third row is absent — this peer has no
#: ``set_expression_evaluator`` — so nothing in this port is dispatch-scoped at all, and the
#: constant reads False because *no entry point takes that shortcut*.
#:
#: One name, two ports, same value, two meanings. That is D13's face amendment one level down
#: — **a claim about a seam names the face it is about** — and it is why the name stays in
#: ``[sdk_surface].substrate`` rather than being read as agreement.
CAPABILITY_CHECK_IS_DISPATCH_SCOPED = False


def assert_not_builtin_override(pattern: str) -> None:
    """§4's override prohibition — **ours to enforce as of v3.29, and it was not before.**

    Through v3.28 §3.5 described this rule as *"a subset of"* core's ``system/*`` reservation
    and told implementers that enforcing that reservation needed *"no separate
    compute-specific guard."* ``ENTITY-CORE-PROTOCOL`` 0.8.2.13 **withdrew the reservation**,
    so v3.29 restates the prohibition on its own basis: it binds **every installation path**
    because it is a cross-peer determinism requirement, not a namespace policy. Two peers
    disagreeing about what ``"add"`` means is an interop failure.

    There is now nothing upstream that refuses a registration at
    ``system/compute/builtins/*`` on our behalf, so this function is the enforcement point,
    and it is called from :func:`~entity_compute.install_compute`'s path rather than left as
    documentation.

    Raises :class:`ValueError`. **Not a ``compute/error``**: this is an INSTALLATION fault in
    the host program, discovered before the peer serves anything, and answering it with a
    value a caller could store would be the §2.4 error type doing a job it is not for.
    """
    relative = pattern
    if pattern.startswith("/"):
        i = pattern.find("/", 1)
        relative = pattern[i + 1:] if i >= 0 else pattern
    if relative == BUILTINS_PREFIX or relative.startswith(BUILTINS_PREFIX + "/"):
        raise ValueError(
            f"EXTENSION-COMPUTE §4: a handler MUST NOT be registered at {pattern}. "
            "The builtin handlers are the cross-peer determinism floor; overriding one makes "
            "two peers disagree about what an operation means. (v3.29 — this rule no longer "
            "delegates to the withdrawn core system/* reservation.)"
        )
